lib/Migration/MultiAccountRepairStep.php

```diff
@@ -32,6 +32,13 @@ public function run(IOutput $output) {
 				$account->setUrl($this->config->getUserUrl($userId));
 				$account->setToken($this->config->getUserToken($userId));
 				$account->setTokenType($this->config->getUserTokenType($userId));
+				if (!$account->getTokenType()) {
+					if ($this->config->hasUserRefreshToken($userId)) {
+						$account->setTokenType('oauth');
+					} else {
+						$account->setTokenType('personal');
+					}
+				}
```

Comment:
```suggestion
				$tokenType = $this->config->getUserTokenType($userId);
				if ($tokenType === '') {
					if ($this->config->hasUserRefreshToken($userId)) {
						$account->setTokenType('oauth');
					} else {
						$account->setTokenType('personal');
					}
				} else {
					$account->setTokenType($tokenType);
				}
```
An explicit check is a lot safer. From looking at your code I'm not even sure it would catch the problem :thinking:

Reply:
Tested it works and have the error message if the fix is not there.
If the config key is not there or is empty, the value would be an empty string, so `!""` is true.

Reply:
But this kind of check is considered risky because it can have unintended behavior. It's always better to use an explicit check for the thing you want to achieve.

Reply:
See https://psalm.dev/docs/running_psalm/issues/RiskyTruthyFalsyComparison/

Reply:
but we want to handle all the cases here, no? All falsy values: null, empty string, everything. If token_type is not there, we put it there.
I'll keep your advice in mind for future.

Reply:
https://github.com/nextcloud/integration_gitlab/blob/2978931c492b764dccc8f6a8c7ae268836cc30ea/lib/Service/ConfigService.php#L61
This is not nullable, so the only invalid state we need to detect if it is an empty string. It doesn't matter if it is an empty string because there is an empty string in the DB or because the value is missing entirely.

Reply:
For extra safety you could also check if the token type matches the two valid values and infer it otherwise. If there ever was a wrong non-empty string saved this would fix it as well.

Reply:
maybe we are over-engineering this small fix :smile: 
> so the only invalid state we need to detect if it is an empty string

and the `if` covers for that

Reply:
Sure, if it works it works :woman_shrugging: